kinto_client/__init__.py

```diff
@@ -166,25 +170,29 @@ def _get_endpoint(self, name, bucket=None, collection=None, id=None):
         }
         return self.endpoints.get(name, **kwargs)
 
-    def _paginated(self, endpoint, records=None, visited=None):
+    def _paginated(self, endpoint, records=None, if_none_match=None,
+                   with_headers=False, **kwargs):
         if records is None:
             records = collections.OrderedDict()
-        if visited is None:
-            visited = set()
-
-        record_resp, headers = self.session.request('get', endpoint)
-        records.update(collections.OrderedDict(
-            [(r['id'], r) for r in record_resp['data']]))
-
-        visited.add(endpoint)
-
-        if 'next-page' in map(str.lower, headers.keys()):
-            # Paginated wants a relative URL, but the returned one is absolute.
-            next_page = headers['Next-Page']
-            # Due to bug mozilla-services/cliquet#366, check for recursion:
-            if next_page not in visited:
-                return self._paginated(next_page, records, visited)
-
+        headers = {}
+        if if_none_match is not None:
+            headers['If-None-Match'] = utils.quote(if_none_match)
+
+        record_resp, headers = self.session.request(
+            'get', endpoint, headers=headers, params=kwargs)
+        if record_resp:
+            records.update(collections.OrderedDict(
+                [(r['id'], r) for r in record_resp['data']]))
+
+            if 'next-page' in map(str.lower, headers.keys()):
+                # Paginated wants a relative URL, but the returned one is
+                # absolute.
+                next_page = headers['Next-Page']
+                return self._paginated(next_page, records,
+                                       with_headers=with_headers,
+                                       if_none_match=if_none_match)
+        if with_headers is True:
```

Comment:
Not really sure about this one: it's a way to return the headers, as they can be useful sometimes.